governance/constitution/actions/default.js

```diff
@@ -0,0 +1,1481 @@
+    checkType(jwk.kty, "string", `${field}.keys[${i}].kty`);
+    checkType(jwk.x5c, "array", `${field}.keys[${i}].x5c`);
+    checkLength(jwk.x5c, 1, null, `${field}.keys[${i}].x5c`);
+    for (const [j, b64der] of jwk.x5c.entries()) {
+      checkType(b64der, "string", `${field}.keys[${i}].x5c[${j}]`);
+      const pem =
+        "-----BEGIN CERTIFICATE-----\n" +
+        b64der +
+        "\n-----END CERTIFICATE-----";
+      checkX509CertBundle(pem, `${field}.keys[${i}].x5c[${j}]`);
+    }
+  }
+}
+
+function checkX509CertBundle(value, field) {
+  if (!ccf.crypto.isValidX509CertBundle(value)) {
+    throw new Error(
+      `${field} must be a valid X509 certificate (bundle) in PEM format`,
+    );
+  }
+}
+
+function invalidateOtherOpenProposals(proposalIdToRetain) {
+  const proposalsMap = ccf.kv["public:ccf.gov.proposals_info"];
+  proposalsMap.forEach((v, k) => {
+    let proposalId = ccf.bufToStr(k);
+    if (proposalId !== proposalIdToRetain) {
+      let info = ccf.bufToJsonCompatible(v);
+      if (info.state === "Open") {
+        info.state = "Dropped";
+        proposalsMap.set(k, ccf.jsonCompatibleToBuf(info));
+      }
+    }
+  });
+}
+
+function setServiceCertificateValidityPeriod(validFrom, validityPeriodDays) {
+  const rawConfig =
+    ccf.kv["public:ccf.gov.service.config"].get(getSingletonKvKey());
+  if (rawConfig === undefined) {
+    throw new Error("Service configuration could not be found");
+  }
+  const serviceConfig = ccf.bufToJsonCompatible(rawConfig);
+
+  const default_validity_period_days = 365;
+  const max_allowed_cert_validity_period_days =
+    serviceConfig.maximum_service_certificate_validity_days ??
+    default_validity_period_days;
+
+  if (
+    validityPeriodDays !== undefined &&
+    validityPeriodDays > max_allowed_cert_validity_period_days
+  ) {
+    throw new Error(
+      `Validity period ${validityPeriodDays} (days) is not allowed: service max allowed is ${max_allowed_cert_validity_period_days} (days)`,
+    );
+  }
+
+  const renewed_service_certificate = ccf.network.generateNetworkCertificate(
+    validFrom,
+    validityPeriodDays ?? max_allowed_cert_validity_period_days,
+  );
+
+  const serviceInfoTable = "public:ccf.gov.service.info";
+  const rawServiceInfo = ccf.kv[serviceInfoTable].get(getSingletonKvKey());
+  if (rawServiceInfo === undefined) {
+    throw new Error("Service info could not be found");
+  }
+  const serviceInfo = ccf.bufToJsonCompatible(rawServiceInfo);
+
+  serviceInfo.cert = renewed_service_certificate;
+  ccf.kv[serviceInfoTable].set(
+    getSingletonKvKey(),
+    ccf.jsonCompatibleToBuf(serviceInfo),
+  );
+}
+
+function setNodeCertificateValidityPeriod(
+  nodeId,
+  nodeInfo,
+  validFrom,
+  validityPeriodDays,
+) {
+  if (nodeInfo.certificate_signing_request === undefined) {
+    throw new Error(`Node ${nodeId} has no certificate signing request`);
+  }
+
+  const rawConfig =
+    ccf.kv["public:ccf.gov.service.config"].get(getSingletonKvKey());
```

Comment:
same here. We could save all settings in the same map.

Reply:
See https://github.com/microsoft/azure-privacy-sandbox-kms/pull/242#issuecomment-2532277071